File: factory_method/dart/factory_method.dart
import 'product_interface.dart';

main() {
  final carTransportation = CarDeliveryTransportFactory();
  final bikeTrasportation = BikeDeliveryTransportFactory();

  carTransportation.transportOrder();
  bikeTrasportation.transportOrder();

  carTransportation.reparingVehicle();
  bikeTrasportation.reparingVehicle();
}

abstract class TransportDeliveryFactory {
  abstract final Vehicle vehicle;

  transportOrder() {
    print("the order is moving in $vehicle ");
  }

  reparingVehicle() {
    print("the $vehicle is in garage ");
  }

  Vehicle createVehicle();
}

class Car extends Vehicle {
  @override
  start() {
    print("start engine");
  }

  @override
  stop() {
    print("stop engine");
  }
}

class Bike extends Vehicle {
  @override
  start() {
    print("start pedal");
  }

  @override
  stop() {
    print("stop pedal");
  }
}

class CarDeliveryTransportFactory extends TransportDeliveryFactory {
  @override
  Vehicle createVehicle() => Car();

  @override
  Vehicle get vehicle => createVehicle();
}
class BikeDeliveryTransportFactory extends TransportDeliveryFactory {
  @override
  Vehicle createVehicle() => Bike();

  @override
  Vehicle get vehicle => createVehicle();
}

File: factory_method/kotlin/FactoryMethod.kt


fun main(){
    val carTransportation = CarDeliveryTransportFactory()
    val bikeTrasportation = BikeDeliveryTransportFactory()

    carTransportation.transportOrder()
    bikeTrasportation.transportOrder()

    carTransportation.reparingVehicle()
    bikeTrasportation.reparingVehicle()
}



abstract class TransportDeliveryFactory{
    abstract val vehicle : Vehicle

    fun transportOrder(){
        println("the order is moving in $vehicle ")
    }
    fun reparingVehicle(){
        println("the $vehicle is in garage ")
    }

    abstract fun createVehicle():Vehicle
}

interface Vehicle {
    fun start()
    fun stop()
}

class Car() : Vehicle{

    override fun start() {
        println("start engine")

     }

    override fun stop() {
        println("stop engine")
     }
     override fun toString() = "Car"
}

class Bike(): Vehicle{

    override fun start() { 
        println("start pedal")
    }

    override fun stop() {
        println("stop pedal")
     }
     override fun toString() = "Bike"
}


class CarDeliveryTransportFactory() : TransportDeliveryFactory(){

    override val vehicle: Vehicle = createVehicle()

    override fun createVehicle(): Vehicle { 
        return Car()
    }

}

class BikeDeliveryTransportFactory(): TransportDeliveryFactory(){

    override val vehicle: Vehicle = createVehicle()

    override fun createVehicle(): Vehicle = Bike()

}